go/controller/internal/controller/helmrepository_controller.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/pluralsh/console/go/controller/internal/credentials"
+	"github.com/pluralsh/console/go/controller/internal/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// HelmRepositoryReconciler reconciles a v1alpha1.HelmRepository object.
+// Implements reconcile.Reconciler and types.Controller.
+type HelmRepositoryReconciler struct {
+	client.Client
+	ConsoleClient    consoleclient.ConsoleClient
+	Scheme           *runtime.Scheme
+	UserGroupCache   cache.UserGroupCache
+	CredentialsCache credentials.NamespaceCredentialsCache
+}
+
+// SetupWithManager is responsible for initializing new reconciler within provided ctrl.Manager.
+func (in *HelmRepositoryReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	mgr.GetLogger().Info("Starting reconciler", "reconciler", "helmrepository_reconciler")
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).                                                             // Requirement for credentials implementation.
+		Watches(&v1alpha1.NamespaceCredentials{}, credentials.OnCredentialsChange(in.Client, new(v1alpha1.HelmRepositoryList))). // Reconcile objects on credentials change.
+		For(&v1alpha1.HelmRepository{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Owns(&corev1.Secret{}, builder.WithPredicates(predicate.ResourceVersionChangedPredicate{})).
+		Complete(in)
+}
+
+// +kubebuilder:rbac:groups=deployments.plural.sh,resources=helmrepositories,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=deployments.plural.sh,resources=helmrepositories/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=deployments.plural.sh,resources=helmrepositories/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the v1alpha1.HelmRepository closer to the desired state
+// and syncs it with the Console API state.
+func (in *HelmRepositoryReconciler) Reconcile(ctx context.Context, req reconcile.Request) (_ reconcile.Result, retErr error) {
+	logger := log.FromContext(ctx)
+
+	helmRepository := new(v1alpha1.HelmRepository)
+	if err := in.Get(ctx, req.NamespacedName, helmRepository); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	utils.MarkCondition(helmRepository.SetCondition, v1alpha1.ReadyConditionType, v1.ConditionFalse, v1alpha1.ReadyConditionReason, "")
+
+	scope, err := NewHelmRepositoryScope(ctx, in.Client, helmRepository)
```

Comment:
If there is no custom logic required for the scope you can use the generic scope (`NewDefaultScope`) same as stack definition and observability provider reconcilers.